File: Java/ArrayListAndStreamPractice/src/WordCount/HighestScoringWord.java
package WordCount;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HighestScoringWord {

    String stringOfWords;

    private int getCharScore(char character) {
        String alphabet = "abcdefghijklmnopqrstuvwxyz";
        return alphabet.indexOf(character) + 1;
    }

    public String getHighestScoringWord() {
        ArrayList<String> listOfWords = new ArrayList<>(List.of(stringOfWords.toLowerCase().split(" ")));
        ArrayList<Integer> listOfWordScores = new ArrayList<Integer>();
        listOfWords.forEach((w) -> listOfWordScores.add(getWordScore(w)));
        Integer maxVal = Collections.max(listOfWordScores);
        return listOfWords.get(listOfWordScores.indexOf(maxVal));
    }

    private int getWordScore(String word) {
        return word.chars()
                .map(x -> getCharScore((char) x))
                .reduce(0, Integer::sum);
    }

    public HighestScoringWord(String stringOfWords) {
        this.stringOfWords = stringOfWords;
    }
}

File: Kotlin/higherOrderArrayMethods/src/highestScoringWord/HighestScoringWord.kt
package highestScoringWord

import calculateSalesAmount.SalesTransaction
import java.math.BigDecimal

class HighestScoringWord(
    private val stringOfWords: String
) {
    private fun getCharScore(char: Char): Int {
        val alphabet = "abcdefghijklmnopqrstuvwxyz"
        return alphabet.indexOf(char) + 1;
    }

    private fun getWordScore(word: String): Int {
        val wordAsArray = word.split("")
        // note: the first element is a comma (quirk of splitting a word in kotlin)
        return wordAsArray.subList(1, wordAsArray.size - 1).sumOf { getCharScore(it.single()) }
        // note: the above is a simplification of:
        //return wordAsArray.subList(1,  wordAsArray.size - 1)
           // .map { getCharScore(it.single()) } // single() converts String to Char
            //.sum() // or can use line below
            //.reduce { acc, value -> acc + value }
    }

    fun getHighestScoringWord(): String {
        val arrayOfWords = stringOfWords.lowercase().split(" ")
        val arrayOfWordScores = ArrayList<Int>()
        arrayOfWords.forEach {
            arrayOfWordScores.add(getWordScore(it))
        }
        val maxWordScoreValue = arrayOfWordScores.max()
        val indexOfMaxValue = arrayOfWordScores.indexOf(maxWordScoreValue)
        return arrayOfWords[indexOfMaxValue]
    }
}